File: frontend/src/objects/problem/ProblemTitle.ts
/** 题目标题品牌类型；用于列表、详情和关联对象展示。 */
export type ProblemTitle = string & { readonly __brand: 'ProblemTitle' }

/** 解析 helper 的返回结果；成功携带品牌值，失败携带可展示错误，不抛异常。 */
type ParseResult<T> = { ok: true; value: T } | { ok: false; error: string }

/** 创建题目标题品牌值；调用前必须完成非空和长度校验。 */
function createProblemTitle(value: string): ProblemTitle {
  /** 注意：这里的 as 只在 parseProblemTitle 校验通过后施加品牌类型。 */
  return value as ProblemTitle
}

/** 将题目标题品牌值还原为字符串；无副作用。 */
export function problemTitleValue(title: ProblemTitle): string {
  return title
}

/** 解析题目标题；去除首尾空白并返回结构化校验结果。 */
export function parseProblemTitle(rawTitle: string): ParseResult<ProblemTitle> {
  const normalized = rawTitle.trim()
  if (!normalized) {
    return { ok: false, error: 'Problem title is required.' }
  }
  if (normalized.length > 120) {
    return { ok: false, error: 'Problem title must be at most 120 characters.' }
  }
  return { ok: true, value: createProblemTitle(normalized) }
}

File: backend/src/main/scala/domains/problem/objects/ProblemTitle.scala
package domains.problem.objects

import io.circe.{Decoder, Encoder}


/** 题目标题；用于列表、详情和提交来源展示。 */
final case class ProblemTitle(value: String)

/** 题目标题的 JSON 编解码与输入校验入口。 */
object ProblemTitle:
  given Encoder[ProblemTitle] = Encoder.encodeString.contramap(_.value)
  given Decoder[ProblemTitle] = Decoder.decodeString.emap(parse)

  /** 规范化并校验标题；空标题和超长标题会被拒绝。 */
  def parse(raw: String): Either[String, ProblemTitle] =
    val normalized = raw.trim
    if normalized.isEmpty then Left("Problem title is required.")
    else if normalized.length > 120 then Left("Problem title must be at most 120 characters.")
    else Right(ProblemTitle(normalized))
